cranelift/docs/isle.md

```diff
@@ -0,0 +1,1306 @@
+```
+
+we eliminate the `Option` on the return type, so the method is instead
+
+```rust
+    trait Context {
+        // ...
+        fn disassemble_b(&mut self, arg0: &U) -> T;
+    }
+```
+
+## ISLE Internals
+
+### Compiler Stages
+
+Some detail and pointers to the compilation stages can be found in the
+[README](../isle/README.md). The sequence starts as any ordinary
+compiler: lexing, parsing, semantic analysis, and generation of an
+IR. The most unique part is the "decision trie generation", which is
+what converts the unordered-list-of-rule representation into something
+that corresponds to the final Rust code's control flow and order of
+matching operations.
+
+### Decision Trie
+
+The heart of the ISLE transformation lies in how the compiler converts
+a list of rules into a scheme to attempt to match rules in some order,
+possibly sharing match operations between similar rules to reduce
+work.
+
+The core data structure we produce is a "decision trie" per internal
+constructor body. This is an intermediate representation of sorts that
+is built from individual-rule IR (LHS + RHS) sequences, and is then
+used to generate Rust source.
+
+The decision trie is, as the name implies, a kind of decision tree, in
+the sense that we start at the root and move down the tree based on
+the result of match operations (each feeding one "decision").
+
+It is a "trie" (which is a kind of tree) because at each level, its
+edges are labeled with match operations; a trie is a tree where one
+input character from an alphabet is used to index children at each
+level.
+
+Each node in the tree is either an internal decision node, or a leaf
+"expression" node (which we reach once we have a successful rule
+match). The "execution semantics" of the trie are
+backtracking-based. We attempt to find some path down the tree through
+edges whose match ops run successfully; when we do this to reach a
+leaf, we have the values generated by all of the match ops, and we can
+execute the sequence of "expression instructions" in the leaf. Each
+rule's left-hand side becomes a series of edges (merged into the
+existing tree as we process rules) and each rule's right-hand side
+becomes one leaf node with expression instructions.
+
+At any point, if a match op does not succeed, we try the next out-edge
+in sequence. If we have tried all out-edges from a decision node and
+none were successful, then we backtrack one level further. Thus, we
+simply perform an in-order tree traversal and find the first
+successful match.
+
+Though this sounds possibly very inefficient if some decision node has
+a high fan-out, in practice it is not because the edges are often
+known to be *mutually exclusive*. The canonical example of this is
+when an enum-typed value is destructured into different variants by
+various edges; we can use a Rust `match` statement in the generated
+source and have `O(1)` (or close to it) cost for the dispatch at this
+level.
```

Comment:
In this case, backtracking is impossible, correct? So nodes can only be used in backtracking if they are not enum destructors?

Reply:
Backtracking (up the decision trie) is still possible if no match arm matches; e.g. one can have

```rust
match value {
    Value::A { ref x, ref y } => { ... return Some(...); }
    Value::B { ref z } => { ... return Some(...); }
    _ => {}
}
None
```

We always add the `_ => {}` catch-all, as we don't do exhaustiveness checking on the match arms. (We silence the unreachable-code warning so that rustc, which does do this, doesn't complain!)